addons/controllers/certificates_test.go

```diff
@@ -0,0 +1,107 @@
+package controllers
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	adminregv1 "k8s.io/api/admissionregistration/v1"
+	cert2 "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"path"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/controllers/certificates"
+)
+
+var _ = Describe("Webhook", func() {
+
+	Context("server's certificate and key", func() {
+		It("should be generated and written to the webhook server CertDir", func() {
+			Eventually(func() string { return webhookServer.CertDir }, "10s").ShouldNot(BeEmpty()) //Until the webhook server has comleted startup CertDir returns ""
+			certFilePath := path.Join(webhookServer.CertDir, webhookServer.CertName)
+			Expect(certFilePath).ShouldNot(BeEmpty())
+			keyFilePath := path.Join(webhookServer.CertDir, webhookServer.KeyName)
+			Expect(keyFilePath).ShouldNot(BeEmpty())
+			Expect(webhookServer).ToNot(BeNil())
+			secret, err := certificates.UpdateTLSSecret(ctx, webhookScrtName, webhookServiceName, webhookServer, addonNamespace)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(secret).NotTo(BeNil())
+			cert, err := cert2.CertsFromFile(certFilePath)
+			Expect(err).ToNot(HaveOccurred())
+			certPEM, err := cert2.EncodeCertificates(cert[0])
+			Expect(err).ToNot(HaveOccurred())
+			Expect(certPEM).To(Equal(secret.Data[certificates.ServerCert]))
+			key, err := keyutil.PrivateKeyFromFile(keyFilePath)
+			Expect(err).ToNot(HaveOccurred())
+			orgKey, err := keyutil.ParsePrivateKeyPEM(secret.Data[certificates.ServerKey])
+			Expect(err).ToNot(HaveOccurred())
+			Expect(key).To(Equal(orgKey))
+		})
+	})
+
+	Context("Defautling and Validating Configurations", func() {
+		It("should be updated with CA bundle ", func() {
```

Comment:
Should we do resource clean up after the test case is executed?

Reply:
Do you mean delete the created configurations?

Reply:
Yes, I assume each test will create it's own configuration? Or you thing other tests could come and add rules here?